packages/edit-post/src/components/header/header-toolbar/index.js

```diff
@@ -89,22 +127,82 @@ function HeaderToolbar() {
 				} }
 				disabled={ ! isInserterEnabled }
 				icon={ plus }
+				/* translators: button label text should, if possible, be under 16
+			characters. */
 				label={ _x(
 					'Add block',
 					'Generic label for block inserter button'
 				) }
-			/>
-			{ isLargeViewport && <ToolbarItem as={ ToolSelector } /> }
-			<ToolbarItem as={ EditorHistoryUndo } />
-			<ToolbarItem as={ EditorHistoryRedo } />
-			<ToolbarItem
-				as={ TableOfContents }
-				hasOutlineItemsDisabled={ isTextModeEnabled }
-			/>
-			<ToolbarItem
-				as={ BlockNavigationDropdown }
-				isDisabled={ isTextModeEnabled }
-			/>
+				showTooltip={ ! showIconLabels }
+			>
+				{ showIconLabels && __( 'Add' ) }
+			</ToolbarItem>
+			{ ( isWideViewport || ! showIconLabels ) && (
+				<>
+					{ isLargeViewport && (
+						<ToolbarItem
+							as={ ToolSelector }
+							showTooltip={ ! showIconLabels }
+							isTertiary={ showIconLabels }
+						/>
+					) }
+					<ToolbarItem
+						as={ EditorHistoryUndo }
+						showTooltip={ ! showIconLabels }
+						isTertiary={ showIconLabels }
+					/>
+					<ToolbarItem
+						as={ EditorHistoryRedo }
+						showTooltip={ ! showIconLabels }
+						isTertiary={ showIconLabels }
+					/>
+					{ overflowItems }
+				</>
+			) }
+			{ ! isWideViewport && ! isSmallViewport && showIconLabels && (
+				<DropdownMenu
+					position="bottom right"
+					label={
+						/* translators: button label text should, if possible, be under 16
+	characters. */
+						__( 'Tools' )
+					}
+				>
+					{ () => (
+						<div className="edit-post-header__dropdown">
+							<MenuGroup label={ __( 'Modes' ) }>
+								<MenuItemsChoice
+									value={
+										isNavigationTool ? 'select' : 'edit'
+									}
+									onSelect={ onSwitchMode }
+									choices={ [
+										{
+											value: 'edit',
+											label: <>{ __( 'Edit' ) }</>,
```

Comment:
Nitpick/Question: Do we need to wrap this in a `Fragment`?

Reply:
Whoops, no 😅